lib/reversetunnel/track/tracker.go

```diff
@@ -0,0 +1,298 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package track
+
+import (
+	"context"
+	//"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/workpool"
+)
+
+type Lease = workpool.Lease
+
+// Key uniquely identifies a reversetunnel endpoint.
+type Key struct {
+	Cluster string
+	Type    string
+	Addr    utils.NetAddr
+}
+
+type Config struct {
+	ProxyExpiry time.Duration
+	TickRate    time.Duration
+}
+
+func (c *Config) SetDefaults() {
+	if c.ProxyExpiry < 1 {
+		c.ProxyExpiry = time.Minute * 3
+	}
+	if c.TickRate < 1 {
+		c.TickRate = time.Second * 30
+	}
+}
+
+// Tracker is a helper for tracking proxies located behind reverse tunnels
+// and triggering agent spawning as needed.
+type Tracker interface {
+	// Start starts tracking for specified key.
+	Start(key Key)
+	// Stop stops tracking for specified key.
+	Stop(key Key)
+	// Acquire grants access to the Acquire channel of the
+	// embedded work group.
+	Acquire() <-chan Lease
+	// MarkSeen starts/refreshes tracking for supplied proxies.
+	MarkSeen(lease Lease, proxies ...string)
+	// WithProxy runs the supplied closure if and only if
+	// no other work is currently being done with the proxy
+	// identified by principals.
+	WithProxy(work func(), lease Lease, principals ...string) (didWork bool)
+	// StopAll permanently deactivates this tracker and cleans
+	// up all background goroutines.
+	StopAll()
+}
+
+type proxyTracker struct {
+	sync.Mutex
+	Config
+	wp     workpool.Pool
+	sets   map[Key]*proxySet
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func New(ctx context.Context, c Config) Tracker {
+	ctx, cancel := context.WithCancel(ctx)
+	c.SetDefaults()
+	t := &proxyTracker{
+		Config: c,
+		wp:     workpool.NewPool(ctx),
+		sets:   make(map[Key]*proxySet),
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	go t.run()
+	return t
+}
+
+func (t *proxyTracker) run() {
+	ticker := time.NewTicker(t.TickRate)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			t.Tick()
+		case <-t.ctx.Done():
+			return
+		}
+	}
+}
+
+func (p *proxyTracker) Acquire() <-chan Lease {
+	return p.wp.Acquire()
+}
+
+func (p *proxyTracker) MarkSeen(lease Lease, proxies ...string) {
+	p.Lock()
+	defer p.Unlock()
+	key := lease.Key().(Key)
+	set, ok := p.sets[key]
+	if !ok {
+		return
+	}
+	t := time.Now()
+	for _, name := range proxies {
+		set.markSeen(t, name)
+	}
+	count := len(set.proxies)
+	if count < 1 {
+		count = 1
+	}
+	p.wp.Set(key, uint64(count))
```

Comment:
Won't the count be picked up in the `tick` function and won't it be more accurate because it will have reduce count by the number of proxies that are expired?

Reply:
Expiry is checked on a pretty long interval (30 seconds, though it is configurable).  If we only updated counts at that point, the AgentPool would be very slow to react to new proxies being available (also, tests would take significantly longer).